### ISSUE TITLE
Build of core fails on Windows 7 with bind errors, address already in use

### ISSUE THREAD

**Issue by cmebarrow** [nuklei-io/nuklei#20]
```
Tests in error:
  shouldHandleConnectAndSendReceiveOfDataWithSingleNuklei(org.kaazing.nuklei.net.TcpManagerTest): test timed out after 1000 milliseconds
  shouldHandleConnectAndReceiveOfDataWithSingleNuklei(org.kaazing.nuklei.net.TcpManagerTest): test timed out after 1000 milliseconds
  shouldBeAbleToCloseConnectionWithNoData(org.kaazing.nuklei.net.TcpManagerTest): test timed out after 1000 milliseconds
  shouldConnectOut(org.kaazing.nuklei.net.TcpManagerTest): Address already in use: bind

Tests run: 52, Failures: 0, Errors: 4, Skipped: 0
```
Console output shows numerous exceptions like the following:
```
Exception in thread "TCP-manager-dedicated" java.lang.RuntimeException: java.lang.IllegalStateException: java.net.BindException: Address alr
eady in use: bind
        at org.kaazing.nuklei.MessagingNukleus.process(MessagingNukleus.java:115)
        at org.kaazing.nuklei.DedicatedNuklei.run(DedicatedNuklei.java:101)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: java.net.BindException: Address already in use: bind
        at org.kaazing.nuklei.net.TcpAcceptor.<init>(TcpAcceptor.java:82)
        at org.kaazing.nuklei.net.TcpManager.commandHandler(TcpManager.java:134)
        at org.kaazing.nuklei.net.TcpManager$$Lambda$53/230835489.accept(Unknown Source)
        at org.kaazing.nuklei.concurrent.MpscArrayBuffer.read(MpscArrayBuffer.java:185)
        at org.kaazing.nuklei.MessagingNukleus.process(MessagingNukleus.java:110)
        ... 2 more
Caused by: java.net.BindException: Address already in use: bind
        at sun.nio.ch.Net.bind0(Native Method)
        at sun.nio.ch.Net.bind(Net.java:436)
        at sun.nio.ch.Net.bind(Net.java:428)
        at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:214)
        at java.nio.channels.ServerSocketChannel.bind(ServerSocketChannel.java:157)
        at org.kaazing.nuklei.net.TcpAcceptor.<init>(TcpAcceptor.java:72)
        ... 6 more
```

**Comment by cmebarrow** [comment]
The methods work when run singly, but not when just two are run, for example if test methods shouldBeAbleToCloseConnectionWithData and shouldBeAbleToCloseConnectionWithNoData are run in one process using command line
mvn test -Dtest=TcpManagerTest#shouldBeAbleToCloseConnection* | tee test.out
the second one to run (shouldBeAbleToCloseConnectionWithData) fails with the bind errors.

**Comment by cmebarrow** [comment]
We hit and fixed a similar issue in Netty back in 2013. Here are the notes from the bug (from Josh):
I think I have this nailed (finally).
sun.nio.ch.SocketChannelImpl does the following:
executes a native preClose function (which only exists on UNIX based OS's).
signals a native thread if needed (test case doesn't do this).
If the ServerSocketChannel has NO registered SelectionKeys than it executes the native close function
It then cancels all SelectionKeys
The problem occurs when there are registered selectionkeys and on Windows. The problem does not occur on UNIX system because of the preClose function. Specifically it does this:
dup2( someOldClosedFD, socketFD )
Consequently since socketFD is open it gets closed as a side affect. Why does it do this on Unix? The comments in sun..nio.chNativeDispatcher says the following:
// Prepare the given fd for closing by duping it to a known internal fd
// that's already closed. This is necessary on some operating systems
// (Solaris and Linux) to prevent fd recycling.
Ok. So here is a code segment of the a failing NIO test:
serverSocketChannel.socket().bind(address);
Selector selector = Selector.open();
SelectionKey key = serverSocketChannel.register(selector, SelectionKey.OP_ACCEPT, null);
serverSocketChannel.close();
Assert.assertTrue("socket is bound", checkIfUnbound(bindURI));
On Windows this fails, on Linux it succeeds (should on Mac as well). But we can work around this. The NIO doc (http://docs.oracle.com/javase/7/docs/api/java/nio/channels/SelectionKey.html) says that canceled selector keys are removed from the selector during the next select operation. It also says (when the channel is closed ... so I'd call this a NIO bug). So the workaround is to do a select after the close (since close cancels it) or to cancel all keys and call select before the close.
In other words this passes on windows:
serverSocketChannel.socket().bind(address);
Selector selector = Selector.open();
SelectionKey key = serverSocketChannel.register(selector, SelectionKey.OP_ACCEPT, null);
serverSocketChannel.close();
selector.select(1);
Assert.assertTrue("socket is bound", checkIfUnbound(bindURI));
as does this:
serverSocketChannel.socket().bind(address);
Selector selector = Selector.open();
SelectionKey key = serverSocketChannel.register(selector, SelectionKey.OP_ACCEPT, null);
key.cancel();
selector.select(1);
serverSocketChannel.close();
Assert.assertTrue("socket is bound", checkIfUnbound(bindURI));

**Comment by tmontgomery** [comment]
Most likely this is due to shutdown of tests not calling `Selector.select` to free the ports.

**Comment by cmebarrow** [comment]
Exactly. I have coded that fix and it resolves the problem. NIO bug?

**Comment by tmontgomery** [comment]
Unfortunately, the JDK team doesn't consider it a bug. It's just how Java is build once, run everywhere.... NOT!

**Comment by cmebarrow** [comment]
Fix has been merged
